### PR TITLE
Nested data structure from bo file

### PR DESCRIPTION
**Pull request by katiebuntic**
Having a start at using dataclasses for creating a nested structure from the BO file

### CODE REVIEW

File: dataclass_function.py
Comment (stefpiatek) [inline dataclass_function.py:81]: maybe `School` would be clearer?

File: dataclass_function.py
Comment (katiebuntic) [inline dataclass_function.py:70]: Need to change this to lowercase

**Review by stefpiatek — COMMENTED** [review]
This looks good! I think we can either use this directly or I wonder if we can use https://pypi.org/project/pandas-dataclasses/ to have a nested pandas structure. Worth having a quick look to see if installing that and we can use pandas

File: dataclass_function.py
Comment (katiebuntic) [reply dataclass_function.py:81]: Oh yes agreed!

Comment (katiebuntic) [reply dataclass_function.py:81]: Great, thanks! I will have a further look at nested pandas structure now

**Comment by stefpiatek** [comment]
👌 Might actually be better to have 3 separate data frames that you can join together, the Teacher data type would have `RREDUserID` as the index, but also have the `school_id` as a column. Then the `School` data type would have `school_id` as an index?

**Comment by katiebuntic** [comment]
> 

Sounds good!

**Comment by katiebuntic** [comment]
@stefpiatek just updated it to the pandas way - is this what you had in mind? I will do some tests with it with the BO file data

File: dataclass_function.py
Comment (stefpiatek) [inline dataclass_function.py:11]: oh what's complex?

**Review by stefpiatek — COMMENTED** [review]
🎉 Yeah perfect. Great, hopefully should be easy to convert from the massive data frame into our nested structure

**Review by stefpiatek — COMMENTED** [review]
Looking good, will wait for approval until we have functions and tests that can convert the input file to the data structure.

If we wanted to be super sure as well, we could have a data type of the original input file, so we're explicit about what we're expecting.

File: dataclass_function.py
Comment (stefpiatek) [inline dataclass_function.py:14]: out of interest does this work from pandas? Would be preferable if it did
```suggestion
    assessi_engtest2: Data[Int32Dtype]
```


 If not then should add a comment the first time so that we know why we're using floats

File: dataclass_function.py
Comment (stefpiatek) [inline dataclass_function.py]: Worth checking in review that we haven't lost any datetimes

File: dataclass_function.py
Comment (stefpiatek) [inline dataclass_function.py:114]: would we need to make the schools unique first?

it may handle it but I imaginged making a new df which is just the school columns, unique and then create all of the schools like this?

You can then remove the school columns for the next step

File: dataclass_function.py
Comment (stefpiatek) [inline dataclass_function.py:125]: then once you got to here, I expect you could do something like passing in all of the rows to the Pupil constructor which harry should be able to help with if it gets really hairy

File: dataclass_function.py
Comment (katiebuntic) [reply dataclass_function.py:114]: Ah yes that would be much nicer, will update this

File: dataclass_function.py
Comment (katiebuntic) [reply dataclass_function.py:14]: This doesn't work :( so will just add a comment on why

**Comment by katiebuntic** [comment]
Have added another file for the Main df `dataclass_function_main.py` which is indexing the school + teacher, then the pupil information is the data. It looks alright but may not be what is needed..

I could not seem to work nesting the three dfs within pandas dataclass (`dataclass_function.py`) - may need some help with this next week if poss @harryjmoss :) 
image

**Comment by harryjmoss** [comment]
@katiebuntic have gone through this and things are looking good generally! I'd stared at the problem for a while thinking about the nested dataframe structure and came up with a few changes in [this pull request](https://github.com/UCL-ARC/rred-reports/pull/21) - feel free to take as many or as few of these are you like. Will also need some tests here as Stef mentioned.

@stefpiatek apols if this wasn't the nested structure we're ultimately going for!

**Comment by katiebuntic** [comment]
@harryjmoss ready for review now :)

File: src/rred_reports/dataclass_function.py
Comment (stefpiatek) [inline src/rred_reports/dataclass_function.py:118]: Instead of having pylint disabling maybe we should edit our config for pylint. Looks like we can do this [for pandas](https://stackoverflow.com/questions/33961756/disabling-pylint-no-member-e1101-error-for-specific-libraries#35106735)  I think we should be able to do this in the `pyproject.toml` file, under its  `[tool.pylint]` section

File: src/rred_reports/dataclass_function.py
Comment (stefpiatek) [inline src/rred_reports/dataclass_function.py:185]: this would be great to have moved to a test, and add the example_dataset. Maybe adding some asserts at the end (doesn't need to be extensive) to make sure that the dimensions the dataframes are right and that a couple of the values are correct.

**Review by stefpiatek — APPROVED** [review]
Looks good, thanks for getting this in 🎉 

I think worth addressing all of the changes in here, happy to chat through if anything is unclear.

**Comment by katiebuntic** [comment]
Hi @harryjmoss and @stefpiatek - I have made a couple of changes for review:
- Logger import added in
- Changed all `floats` to `pd.Int32Dtype` in the pandas dataclasses
- Added in the .new when creating the dataframe
- Tests at the end to check all works

There were still some issues with the pylint so I also added in `# pylint: disable=E1121' when creating the dfs - I added in a line to the `pyproject.toml` for this but may need a check as didn't seem to help?

Also, the `example_dataset.xlxs` didn't work when it was in the `test` folder so I have kept it in the main for now

Thank you! :)

**Comment by stefpiatek** [comment]
yeah annoying that the pylint config didn't work but hey ho, I'll create an issue for it. 
Cool, I think merge it in 👏 , I might have some time today between things to set up some quick tests for it and get the file working from the test dir

**Pull request merged**
